spyder/plugins/ipythonconsole/widgets/main_widget.py

```diff
@@ -1571,23 +1591,81 @@ def create_client_for_kernel(self):
             self._create_client_for_kernel(connection_file, hostname, sshkey,
                                            password)
 
-    def connect_client_to_kernel(self, client, is_cython=False,
-                                 is_pylab=False, is_sympy=False):
-        """Connect a client to its kernel."""
-        connection_file = client.connection_file
-        stderr_handle = (
-            None if self._test_no_stderr else client.stderr_obj.handle)
-        stdout_handle = (
-            None if self._test_no_stderr else client.stdout_obj.handle)
+    def get_new_kernel(self, is_cython=False, is_pylab=False,
+                       is_sympy=False, std_dir=None):
+        """Get a new kernel, and cache one for next time."""
+        # Cache another kernel for next time.
+        kernel_spec = self.create_kernel_spec(
+            is_cython=is_cython,
+            is_pylab=is_pylab,
+            is_sympy=is_sympy)
+
+        new_kernel = self.create_new_kernel(kernel_spec, std_dir)
+        if new_kernel[2] is None:
+            # error
+            return new_kernel
+
+        # Check cached kernel has the same configuration as is being asked
+        if self._cached_kernel is not None:
+            (cached_spec,
+             cached_env,
+             cached_argv,
+             cached_dir,
+             cached_kernel) = self._cached_kernel
```

Comment:
This is confusing: having an attribute called `self._cached_kernel` which contains a `cached_kernel` but it's different from that cached kernel.

So, I suggest to rename `self._cached_kernel` to `self._cached_kernel_properties` to make the difference between the two.